server/app/auth/ApiAuthenticator.java

```diff
@@ -74,18 +78,20 @@ public void validate(Credentials rawCredentials, WebContext context, SessionStor
     // Cache the API key for quick lookup in the controller, also for subsequent requests.
     // We intentionally cache the empty optional rather than throwing here so that subsequent
     // requests with the invalid key do not put pressure on the database.
-    ApiKey apiKey =
-        apiKeyService
-            .get()
-            .findByKeyIdWithCache(keyId)
-            .orElseThrow(() -> new BadCredentialsException("API key does not exist"));
+    Optional<ApiKey> maybeApiKey = apiKeyService.get().findByKeyIdWithCache(keyId);
+
+    if (!maybeApiKey.isPresent()) {
+      unauthorized(context, "API key does not exist: " + keyId);
```

Comment:
High-level question on the design. API keys are meant to only be shown once to the API user. Consider the situation where there was a bug in findByKeyIdWithCache where the API key *did* exist, was valid, but wasn't being returned by the method. Here, someone with server logs access would be able to see the actual key. If we then fixed the bug, the key would effectively be "leaked".

This may not be a concern due to the presumable level of access someone with server logs access has, but I wanted to explicitly call it out.

My whole concern could be avoided if we didn't print the key ID here. In my opinion, it's not necessary information for debugging, since there's no real action for someone observing logs to take based on this.

Reply:
Same for the below updates as well.